x/staking/keeper/msg_server.go

```diff
@@ -373,3 +377,193 @@ func (k msgServer) Undelegate(goCtx context.Context, msg *types.MsgUndelegate) (
 		CompletionTime: completionTime,
 	}, nil
 }
+
+func getShareTokenDenom(validatorAddress string, tokenizeShareRecordId uint64) string {
+	return validatorAddress + strconv.Itoa(int(tokenizeShareRecordId))
+}
+
+func (k msgServer) TokenizeShares(goCtx context.Context, msg *types.MsgTokenizeShares) (*types.MsgTokenizeSharesResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	valAddr, valErr := sdk.ValAddressFromBech32(msg.ValidatorAddress)
+	if valErr != nil {
+		return nil, valErr
+	}
+	validator, found := k.GetValidator(ctx, valAddr)
+	if !found {
+		return nil, types.ErrNoValidatorFound
+	}
+
+	_ = validator
+
+	delegatorAddress, err := sdk.AccAddressFromBech32(msg.DelegatorAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	delegation, found := k.GetDelegation(ctx, delegatorAddress, valAddr)
+	if !found {
+		return nil, types.ErrNoDelegatorForAddress
+	}
+
+	delegationAmount := validator.Tokens.ToDec().Mul(delegation.GetShares()).Quo(validator.DelegatorShares)
+	if msg.Amount.Amount.GT(sdk.Int(delegationAmount)) {
+		return nil, types.ErrNotEnoughDelegationShares
+	}
+
+	acc := k.authKeeper.GetAccount(ctx, delegatorAddress)
+	if acc != nil {
+		acc, ok := acc.(vesting.VestingAccount)
+		if ok {
+			// if account is a vesting account, it checks if free delegation (non-vesting delegation) is not exceeding
+			// the tokenize share amount and execute further tokenize share process
+			// tokenize share is reducing unlocked tokens delegation from the vesting account and further process
+			// is not causing issues
+			delFree := acc.GetDelegatedFree().AmountOf(msg.Amount.Denom)
+			if delFree.LT(msg.Amount.Amount) {
+				return nil, types.ErrExceedingFreeVestingDelegations
+			}
+		}
+	}
+
+	recordId := k.GetLastTokenizeShareRecordId(ctx) + 1
+	shareTokenDenom := getShareTokenDenom(msg.ValidatorAddress, recordId)
+	record := types.TokenizeShareRecord{
+		Id:              recordId,
+		Owner:           msg.TokenizedShareOwner,
+		ShareTokenDenom: shareTokenDenom,
+		ModuleAccount:   fmt.Sprintf("tokenizeshare_%d", recordId),
+		Validator:       msg.ValidatorAddress,
+	}
+
+	shareToken := sdk.NewCoin(shareTokenDenom, msg.Amount.Amount)
+
+	err = k.bankKeeper.MintCoins(ctx, minttypes.ModuleName, sdk.Coins{shareToken})
+	if err != nil {
+		return nil, err
+	}
+
+	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, minttypes.ModuleName, delegatorAddress, sdk.Coins{shareToken})
+	if err != nil {
+		return nil, err
+	}
+
+	shares, err := k.ValidateUnbondAmount(
+		ctx, delegatorAddress, valAddr, msg.Amount.Amount,
+	)
+
+	returnAmount, err := k.Unbond(ctx, delegatorAddress, valAddr, shares)
+	if err != nil {
+		return nil, err
+	}
+
+	if validator.IsBonded() {
+		k.bondedTokensToNotBonded(ctx, returnAmount)
+	}
+
+	// create reward ownership record
+	k.AddTokenizeShareRecord(ctx, record)
+
+	_, err = k.Keeper.Delegate(ctx, record.GetModuleAddress(), msg.Amount.Amount, sdkstaking.Unbonded, validator, false)
```

Comment:
if the type of `acc` is a vesting account, we need to call `TrackDelegation` similar to here
https://github.com/cosmos/cosmos-sdk/blob/8b78406bd3941fc0b98683bac06de55898859486/x/bank/keeper/keeper.go#L476-L477